influxdb3_load_generator/src/specification.rs

```diff
@@ -0,0 +1,112 @@
+
+    pub fn to_json_string_pretty(&self) -> Result<String, anyhow::Error> {
+        let res = serde_json::to_string_pretty(&self).context("failed to encode json to string")?;
+        Ok(res)
+    }
+}
+
+/// Specification for a measurement to be generated
+#[derive(Debug, Deserialize, Serialize)]
+pub struct MeasurementSpec {
+    /// The name of the measurement
+    pub name: String,
+    /// The tags to be generated for each line
+    pub tags: Vec<TagSpec>,
+    /// The fields to be generated for each line
+    pub fields: Vec<FieldSpec>,
+    /// Create this many copies of this measurement in each sample. The copy number will be
+    /// appended to the measurement name to uniquely identify it./
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub copies: Option<usize>,
+    /// If this measurement has tags with cardinality, this is the number of lines that will
+    /// be output per sample go through the cardinality. If not specified, all unique values
+    /// will be used. This number must be <= cardinality of the highest cardinality tag.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub lines_per_sample: Option<usize>,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct TagSpec {
+    /// the key/name of the tag
+    pub key: String,
+
+    /// have this many copies of this tag in the measurement. Random values will be generated
+    /// independently (i.e. copies won't share the same random value). Will add the copy number to
+    /// the key of the tag to uniquely identify it.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub copies: Option<usize>,
+    /// if set, appends the copy id of the tag to the value of the tag
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub append_copy_id: Option<bool>,
+
+    /// output this string value for every line this tag is present
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub value: Option<String>,
+
+    /// if set, appends the writer id to the value of the tag
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub append_writer_id: Option<bool>,
+
+    /// will add a number to the value of the tag, with this number of unique values
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub cardinality: Option<usize>,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct FieldSpec {
+    // These options apply to any type of field
+    /// the key/name of the field
+    pub key: String,
+    /// have this many copies of this field in the measurement. Random values will be generated
+    /// independently (i.e. copies won't share the same random value). Will add the copy number to
+    /// the key of the field to uniquely identify it.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub copies: Option<usize>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub null_probability: Option<f64>,
+
+    // These options apply to specific types of fields and are mutually exclusive
```

Comment:
This may be a case for an enum paired with `#[serde(flatten)]`. For example,
```rust
#[derive(Debug, Deserialize, Serialize)]
pub struct FieldSpec {
    // [...]
    #[serde(flatten)]
    pub field: FieldKind,
}

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum FieldKind {
    Bool(bool),
    String(String),
    StringRandom(usize),
    Integer(i64),
    IntegerRange(i64, i64),
    // ...
}
```
That will serialize from / deserialize to the same structure that you have here, but give you the ability to use `match` when handling the spec in code.